File: julia/robot-simulator/robot-simulator.jl
@enum Orientation NORTH EAST SOUTH WEST

struct Point{T<:Real}
    x::T
    y::T
end
Base.:+(p::Point, q::Point) = Point(p.x + q.x, p.y + q.y)

const advance_table = Dict(
    NORTH => Point{Int}(0, 1),
    EAST  => Point{Int}(1, 0),
    SOUTH => Point{Int}(0, -1),
    WEST  => Point{Int}(-1, 0)
)

mutable struct Robot
    position::Point{Int}
    orientation::Orientation
    Robot(p::Tuple{Int,Int}, o::Orientation) = new(Point(p[1], p[2]), o)
end

position(bot::Robot) = bot.position
heading(bot::Robot) = bot.orientation

function turn_right!(bot::Robot)::Robot
    bot.orientation = Orientation(mod(Int(bot.orientation) + 1, 4))
    bot
end
function turn_left!(bot::Robot)::Robot
    bot.orientation = Orientation(mod(Int(bot.orientation) - 1, 4))
    bot
end

function advance!(bot::Robot)::Robot
    bot.position += advance_table[bot.orientation]
    bot
end
const move_table = Dict(
            'A' => advance!,
            'R' => turn_right!,
            'L' => turn_left!
)
function move!(bot::Robot, instruction::String)::Robot
    for cmd in instruction
        move_table[cmd](bot)
    end
    bot
end

File: lua/robot-simulator/robot-simulator.lua
local directions = {
  north = { L = 'west', R = 'east', y = 1 },
  east = { L = 'north', R = 'south', x = 1 },
  south = { L = 'east', R = 'west', y = -1 },
  west = { L = 'south', R = 'north', x = -1 }
  }

return function(p)
  return {
    x = p.x,
    y = p.y,
    heading = p.heading,
    move = function (self, command)
        for i = 1, #command do
            local c = command:sub(i,i)
            if c == 'R' or c == 'L' then self.heading = directions[self.heading][c]
            elseif c == 'A' then
                self.x = self.x + (directions[self.heading].x or 0)
                self.y = self.y + (directions[self.heading].y or 0)
            else
                assert(false, 'unknown command')
            end
        end
    end
  }
end
